File: Sources/YTools/Modules/Calculator/CalculatorModule.swift
import Foundation
import YToolsCore
import YToolsModuleKit

struct CalculatorModule: YToolsModule {
    let descriptor = ModuleDescriptor(id: "calculator", name: "计算器")
    private let calculator = ExpressionCalculator()

    func search(_ request: ModuleSearchRequest) async throws -> [LauncherResult] {
        let rawExpression = request.query.trimmingCharacters(in: .whitespacesAndNewlines)
        let shouldContinue = rawExpression.hasSuffix("=")
        let expression = shouldContinue
            ? String(rawExpression.dropLast()).trimmingCharacters(in: .whitespacesAndNewlines)
            : rawExpression
        guard !expression.isEmpty,
              expression.rangeOfCharacter(from: CharacterSet(charactersIn: "+-*/%^×÷")) != nil
                || expression.rangeOfCharacter(from: .letters) != nil,
              let value = try? calculator.evaluate(expression) else {
            return []
        }

        let result = calculator.format(value)
        return [LauncherResult(
            id: "calculator:\(expression)",
            moduleID: descriptor.id,
            title: result,
            subtitle: shouldContinue ? "\(expression)  ·  回车回填并继续计算" : "\(expression)  ·  回车复制结果",
            icon: .system("function"),
            score: 1_000,
            action: shouldContinue ? .navigate(result) : .copy(result)
        )]
    }
}

File: src/YTools.Windows/Services/Modules/CalculatorModule.cs
using YTools.Core;
using YTools.ModuleKit;

namespace YTools.Services.Modules;

public sealed class CalculatorModule : IYToolsModule
{
    public ModuleDescriptor Descriptor { get; } = new("calculator", "计算器");

    public Task<IReadOnlyList<LauncherResult>> SearchAsync(ModuleSearchRequest request)
    {
        var rawExpression = request.Query.Trim();
        var shouldContinue = rawExpression.EndsWith('=');
        var expression = shouldContinue
            ? rawExpression[..^1].Trim()
            : rawExpression;
        var hasOperator = expression.IndexOfAny(['+', '-', '*', '/', '%', '^', '×', '÷']) >= 0;
        var hasLetters = expression.Any(char.IsLetter);
        if (string.IsNullOrEmpty(expression) || (!hasOperator && !hasLetters))
        {
            return Task.FromResult<IReadOnlyList<LauncherResult>>([]);
        }

        double value;
        try
        {
            value = ExpressionCalculator.Evaluate(expression);
        }
        catch
        {
            return Task.FromResult<IReadOnlyList<LauncherResult>>([]);
        }

        var result = ExpressionCalculator.Format(value);
        return Task.FromResult<IReadOnlyList<LauncherResult>>(
        [
            new LauncherResult(
                $"calculator:{expression}",
                Descriptor.Id,
                result,
                shouldContinue ? $"{expression}  ·  回车回填并继续计算" : $"{expression}  ·  回车复制结果",
                new ResultIcon.System("function"),
                1_000,
                shouldContinue ? new ResultAction.Navigate(result) : new ResultAction.Copy(result))
        ]);
    }
}
